mipidsi/src/interface/parallel.rs

```diff
@@ -0,0 +1,260 @@
+                let last = self.last.take();
+
+                let changed = match last {
+                    Some(old_value) => value ^ old_value,
+                    None => !0, // all ones, this ensures that we will update all the pins
+                };
+
+                $(
+                    let mask = 1 << $x;
+                    if changed & mask != 0 {
+                        if value & mask != 0 {
+                            self.pins.$x.set_high()
+                        } else {
+                            self.pins.$x.set_low()
+                        }
+                        ?;
+                    }
+                )*
+
+                self.last = Some(value);
+                Ok(())
+            }
+        }
+
+        impl<$($PX, )*> From<($($PX, )*)>
+            for $GenericxBitBus<$($PX, )*>
+        where
+            $($PX: OutputPin, )*
+        {
+            fn from(pins: ($($PX, )*)) -> Self {
+                Self::new(pins)
+            }
+        }
+    };
+}
+
+generic_bus! {
+    Generic8BitBus {
+        type Word = u8;
+        Pins {
+            P0 => 0,
+            P1 => 1,
+            P2 => 2,
+            P3 => 3,
+            P4 => 4,
+            P5 => 5,
+            P6 => 6,
+            P7 => 7,
+        }
+    }
+}
+
+generic_bus! {
+    Generic16BitBus {
+        type Word = u16;
+        Pins {
+            P0 => 0,
+            P1 => 1,
+            P2 => 2,
+            P3 => 3,
+            P4 => 4,
+            P5 => 5,
+            P6 => 6,
+            P7 => 7,
+            P8 => 8,
+            P9 => 9,
+            P10 => 10,
+            P11 => 11,
+            P12 => 12,
+            P13 => 13,
+            P14 => 14,
+            P15 => 15,
+        }
+    }
+}
+
+/// Parallel interface error
+#[derive(Clone, Copy, Debug)]
+pub enum ParallelError<BUS, DC, WR> {
+    /// Bus error
+    Bus(BUS),
+    /// Data/command pin error
+    Dc(DC),
+    /// Write pin error
+    Wr(WR),
+}
+
+/// Parallel communication interface
+///
+/// This interface implements a "8080" style write-only display interface using any
+/// [OutputBus] implementation as well as one
+/// `OutputPin` for the data/command selection and one `OutputPin` for the write-enable flag.
+///
+/// All pins in the data bus are supposed to be high-active. High for the D/C pin meaning "data" and the
+/// write-enable being pulled low before the setting of the bits and supposed to be sampled at a
+/// low to high edge.
+pub struct ParallelInterface<BUS, DC, WR> {
+    bus: BUS,
+    dc: DC,
+    wr: WR,
+}
+
+impl<BUS, DC, WR> ParallelInterface<BUS, DC, WR>
+where
+    BUS: OutputBus,
+    BUS::Word: From<u8> + Eq,
+    DC: OutputPin,
+    WR: OutputPin,
+{
+    /// Create new parallel GPIO interface for communication with a display driver
+    pub fn new(bus: BUS, dc: DC, wr: WR) -> Self {
+        Self { bus, dc, wr }
+    }
```

Comment:
I don't believe that is a problem in this case. Unlike CS, WR only cares about the rising *edge*. For each word we set it low, which may or may not be a falling edge, then we set it high, which is definitely a rising edge.

Reply:
OK, that's a valid point. I prefer the pins to be in a known state, but because we don't control CS driving WR high in `new` could even cause an unwanted rising edge on WR. I think we should revisit this later and add optional support for driving CS. If we decide to add read support the correct initial state will become more relevant because the display controller uses the falling edge of RD to put data on the bus.

Reply:
That reminds me, leaving CS and RD floating is a pretty big footgun. Maybe we should require both of them by default. I'll see if I can make the ergonomics work

Reply:
That's going to require more effort, let's leave it for later